### PR TITLE
refactor: speed up mem with notrace

### PR DESCRIPTION
**Pull request by rgrinberg**

### CODE REVIEW

File: lib/hamt.ml [hunk 1]
```diff
@@ -486,33 +486,45 @@ module Make (Config : CONFIG) (Key : Hashtbl.HashedType) :
...
-            let idx = ctpop (bitmap land pred bit) in
-            find (shift + shift_step) hash key base.(idx)
-      | ArrayNode (_, children) ->
-          let child = children.(hash_fragment shift hash) in
-          if is_empty child then raise Not_found
-          else find (shift + shift_step) hash key child
-    in
-    fun key -> find 0 (hash key) key
+  let rec assoc_notrace k = function
+    | [] -> raise Not_found
```
Comment (thizanne) [inline lib/hamt.ml:490]: raise_notrace ?

**Review by thizanne — COMMENTED** [review]
Ι'm not really up to date on this. Would it be faster to implement it with option and wrap it on the exception-raising function, or is the opposite better?

**Comment by rgrinberg** [comment]
> Would it be faster to implement it with option and wrap it on the exception-raising function, or is the opposite better?

I'd say the trade off is as follows:
* If the element is present, exception raising is faster
* If the element is absent, returning None is faster - unless you're using raise_notrace.

**Comment by thizanne** [comment]
That was my intuition, thanks for the confirmation. Let's keep raise_notrace then indeed.

**Pull request merged**